### PR TITLE
[Ready] SoundComponent

### PR DESCRIPTION
**Pull request by Zumorica**
Requires https://github.com/space-wizards/space-station-14/pull/768

- [x] Play sounds
- [x] SoundSchedules actually work
- [x] Send sound to specific users
- [x] Make existing components use SoundComponent
- [x] Add ScheduledSounds from prototypes
- [x] Add Play methods equivalent to those of AudioSystem.
- [x] Document most code.

### CODE REVIEW

File: Content.Shared/GameObjects/Components/Sound/SharedSoundComponent.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        /// </summary>
+        Global,
+        /// <summary>
+        /// Sound is static in a given position.
+        /// </summary>
+        Positional,
+    }
+
+    [Serializable, NetSerializable]
+    public class SoundSchedule
```
Comment (RemieRichards) [inline Content.Shared/GameObjects/Components/Sound/SharedSoundComponent.cs]: I don't know about others (@PJB3005, @Acruid, @clusterfack) but I'd prefer `ScheduledSound` to `SoundSchedule`
as this class represents one sound that has been scheduled, not a whole schedule of sounds.

Comment (Zumorica) [reply Content.Shared/GameObjects/Components/Sound/SharedSoundComponent.cs]: Hmm, you're right. ScheduledSound sounds better, and I think it's easier to understand. I'll change it

File: Content.Server/Content.Server.csproj [hunk 2]
```diff
@@ -158,6 +159,10 @@
       <Project>{59250baf-0000-0000-0000-000000000000}</Project>
       <Name>Lidgren.Network</Name>
     </ProjectReference>
+    <ProjectReference Include="..\engine\SS14.Client\SS14.Client.csproj">
```
Comment (PJB3005) [inline Content.Server/Content.Server.csproj]: Uhm? You're making the content server reference the client here.

File: Content.Shared/GameObjects/Components/Sound/SharedSoundComponent.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            Directed = true;
+        }
+    }
+
+    public enum SoundType
+    {
+        /// <summary>
+        /// Sound follows the entity.
+        /// </summary>
+        Normal,
```
Comment (PJB3005) [inline Content.Shared/GameObjects/Components/Sound/SharedSoundComponent.cs]: Rename this to `Entity` I'd say.

File: Content.Server/Content.Server.csproj (hunk 2 above)
Comment (Zumorica) [reply Content.Server/Content.Server.csproj]: Whoops, this was definitely accidental... I didn't notice.

File: Content.Shared/GameObjects/Components/Sound/SharedSoundComponent.cs (hunk 3 above)
Comment (Zumorica) [reply Content.Shared/GameObjects/Components/Sound/SharedSoundComponent.cs]: Yeah, Entity seems more accurate than Normal. I'll change it.

File: Content.Client/GameObjects/Components/Sound/SoundComponent.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+using SS14.Shared.Serialization;
+using SS14.Shared.Timers;
+
+namespace Content.Client.GameObjects.Components.Sound
+{
+    public class SoundComponent : SharedSoundComponent
+    {
+        private readonly List<ScheduledSound> _schedules = new List<ScheduledSound>();
+        private AudioSystem _audioSystem;
+        private readonly Random Random = new Random();
```
[the remark this answers is not in the document]
Comment (Zumorica) [reply Content.Client/GameObjects/Components/Sound/SoundComponent.cs]: Done. I also made it so Random gets initialized when the first audio plays...

File: Content.Client/GameObjects/Components/Sound/SoundComponent.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            _schedules.Clear();
+        }
+
+        public override void StopScheduledSound(string filename)
+        {
+            foreach (var schedule in _schedules)
+            {
+                if (schedule.Filename != filename) continue;
+                schedule.Play = false;
+                _schedules.Remove(schedule);
```
[the remark this answers is not in the document]
Comment (Zumorica) [reply Content.Client/GameObjects/Components/Sound/SoundComponent.cs]: Indeed, I must have overlooked that... Fixed it already.

**Comment by Zumorica** [comment]
I made it like this so one can make ambient noises with prototypes, since those will most likely be "global" sounds, and not follow the entity. I made an "ambientsound" prototype before, maybe I could add it here too.

**Comment by PJB3005** [comment]
I don't think entities should be managing something that isn't related to themselves. It's fine for code to invoke directly into the audio system for "fire and forget" type code IMO.

Really I want a sound component for ability to stop/start sounds and thing such as ambience, mostly.

**Comment by Zumorica** [comment]
Alright, I made it so you can't play sounds following other entities. Please tell me if you think I should change something else!

**Comment by PJB3005** [comment]
I still do think playing sounds at other positions and playing it globally should not be part of this sound component.

**Comment by Zumorica** [comment]
I can remove the ability to play sounds at other positions, but playing sounds globally is required for most ambience sounds.

**Comment by PJB3005** [comment]
Ambient sounds played by entities should always be local to the entity though. Things like generator hums. Playing ambience globally should be a completely different system IMO.

**Comment by Zumorica** [comment]
Ah, I see now! I understood "ambience sounds" as something similar to the ambience sounds that play on SS13 from time to time. In that case, tomorrow I will remove all SoundTypes except for Entity.

**Comment by Zumorica** [comment]
I made the changes. Should be good now, I think.

**Pull request merged**